torch/_inductor/kernel/flex_decoding.py

```diff
@@ -0,0 +1,554 @@
+            # This fails with Triton IR error Assertion `floatAttr.getType() == eltType && "expected float attribute type to equal element type"'. Use tl.dot to access tensor core instead. 
+        
+        # ~~~~~~~~~~~~~~~~~~~ Apply score modification  ~~~~~~~~~~~~~~~~~~~
+        m = offs_m[:, None]
+        n = start_n + offs_n[None, :]
+        {{ modification(
+            subgraph_number=0,
+            output_name="post_mod_scores",
+            score="qk",
+            b="off_hz // H",
+            h="off_hz % H",
+            m="m",
+            n="n",
+            out="qk"
+        ) | indent_except_first(2) }}
+        # TODO: In the case that score_mod is linear, this can be LICMed
+        if not SCORE_MOD_IS_LINEAR:
+            post_mod_scores *= 1.44269504
+        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+        # -- compute scaling constant ---
+        row_max = tl.max(post_mod_scores, 1)
+        m_i_new = tl.maximum(m_i, row_max)
+
+        alpha = tl.math.exp2(m_i - m_i_new)
+        p = tl.math.exp2(post_mod_scores - m_i_new[:, None])
+        if not ROWS_GUARANTEED_SAFE:
+            masked_out_rows = (m_i_new == float("-inf"))
+            alpha = tl.where(masked_out_rows, 0, alpha)
+            p = tl.where(masked_out_rows[:, None], 0, p)
+
+        # -- scale and update acc --
+        acc *= alpha[:, None]
+        if BLOCK_M >= 16:
+            acc = tl.dot(p.to(MATMUL_PRECISION), v.to(MATMUL_PRECISION), acc=acc)
+        else:
+            acc += tl.sum(p.to(MATMUL_PRECISION)[:, :, None] * v.to(MATMUL_PRECISION), axis=-2).to(tl.float32)
+
+        # -- update m_i and l_i --
+        l_i = l_i * alpha + tl.sum(p, 1)
+        m_i = m_i_new
+        # update pointers
+        K_block_ptr = tl.advance(K_block_ptr, (0, BLOCK_N))
+        V_block_ptr = tl.advance(V_block_ptr, (BLOCK_N, 0))
+
+    # Store output, logsumexp and rowmax for cross CTA reduction. (all in float32, even when input data are in fp16)
+    tl.store(M_block_ptr, m_i[None, :], boundary_check=(0, 1))
+    tl.store(L_block_ptr, l_i[None, :], boundary_check=(0, 1))
+
+    # -- store output
+    idx_z = off_hz // H
+    idx_h = off_hz % H
+    idx_t = off_t
+    idx_m = offs_m[:, None]
+    idx_d = offs_d[None, :]
+    # TODO generalize and add proper mask support
+    mask = (idx_m < Q_CTX) & (idx_d != -1)
+    {{store_output(("idx_z", "idx_h", "idx_t", "idx_m", "idx_d"), "acc", "mask")}} 
+ """,
+)
+
+
+def flex_decoding_reduction_grid(batch_size, num_heads, split_k, m_model, D, meta):
+    """How is this kernel parallelized?
+    We create a grid of (batch_size * num_heads, ceil_div(n_querys, MODEL_M), ceil_div(D, MODEL_D))
+    Each block is responsible for calculating its own blocks of accumulated output
+    """
+    import triton
+
+    return (batch_size * num_heads, triton.cdiv(m_model, meta["BLOCK_M"]), triton.cdiv(D, meta["BLOCK_D"]))
+
+
+flex_decoding_reduction_template = TritonTemplate(
```

Comment:
I don't think you need a template for this - you can just lower it as a normal reduction I think.